### PR TITLE
Add Writer implementation

### PR DESCRIPTION
**Pull request by gege251**
Add a simple interpreter for Writer effects.
Writer effects in the fake PAB will immediately be sent via websocket to the client in a NewObservableState message, and will not be stored in the PAB.
Another difference is that `w` in `Writer w` must have ToJSON instace, while the Monoid instance is not needed.

### CODE REVIEW

File: test/Spec/MockContract.hs [hunk 1]
```diff
@@ -159,46 +175,56 @@ instance Default ContractEnvironment where
     ContractEnvironment
       { cePABConfig = def {pcNetwork = Mainnet}
       , ceContractInstanceId = ContractInstanceId UUID.nil
+      , ceContractState = ContractState <$> unsafePerformIO $ newTVarIO Map.empty
```
Comment (samuelWilliams99) [inline test/Spec/MockContract.hs:178]: This is a crime but I'll allow it

**Review by samuelWilliams99 — COMMENTED** [review]
Perhaps I'm misunderstanding, but by the looks of it, the behaviour of writers that make them a Monoid has been lost here. Since - most of the time - we use `Last`.
However, it might be worth considering the work required to make the writer act this way - if we end up needing this PAB for another project.
Could you give an estimate as to how much work this would be?

File: test/Spec/MockContract.hs (hunk 1 above)
Comment (gege251) [reply test/Spec/MockContract.hs:178]: Yes, I know, I just needed to initialise this for tests, hoping that this is a justifiable use case.

**Comment by gege251** [comment]
@samuelWilliams99 I am not 100% sure that this would behave the same way as the real PAB, but there is no problem with using a Monoid instance, as long as it also an instance of ToJSON. Just in case, I modified the test to use a `Last String`.

**Comment by samuelWilliams99** [comment]
Oh - I was mostly talking about the behaviour Contracts have where the new state is _combined_ with the previous state, so if it were an array, and you tell another array, it'd concat the arrays together. We use `Last` to avoid this behaviour when we don't need it, but it can sometimes be useful.
But again, not super useful for the cards bridge, so unsure if its worth it. If you had to modify the writer effect to now _store_ the current state for combination with a new state, how much extra work would this be?

**Comment by samuelWilliams99** [comment]
Also, its not passing build unfortunately.

**Comment by gege251** [comment]
Oh I understand, yeah, we need to store the writer log then, but I think that could be achievable by modifying the ContractState STM a bit. I'll try implementing it.

**Comment by samuelWilliams99** [comment]
Yeah okay, sounds good! Are you planning to do that in a separate PR?

**Comment by gege251** [comment]
@samuelWilliams99 I implemented writer monoid storage, using STM. I though it would be easy, but it was a bit of a rabbit hole. Most of all, making the contract state map homogeneous was a bit difficult.
Also PabEffect is now parameterized by `w`, and for some reason it has to be applied to each effect call (like `utxosAt @w pkh`) which clutters the code a bit, but I couldn't find a way around it. 
Tests do not cover the server layer, so I had to do some manual testing, but it is looking good.

**Comment by samuelWilliams99** [comment]
Just changed your base following merges of other branches, seems we've got merge conflicts now

File: src/MLabsPAB/Effects.hs [hunk 2]
```diff
@@ -128,4 +133,87 @@ callRemoteCommand ipAddr ShellArgs {cmdName, cmdArgs, cmdOutParser} =
 quotes :: Text -> Text
 quotes str = "\"" <> str <> "\""
 
-makeEffect ''PABEffect
+-- Couldn't use the template haskell makeEffect here, because it caused an OverlappingInstances problem.
+-- For some reason, we need to manually propagate the @w@ type variable to @send@
```
Comment (samuelWilliams99) [inline src/MLabsPAB/Effects.hs:137]: Well that is unfortunate :/

**Review by samuelWilliams99 — APPROVED** [review]
Solid work, as always :)

**Pull request merged**